ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,938 @@ PHP_FUNCTION(bcscale)
+{
+	bcmath_number_obj_t *intern = zend_object_alloc(sizeof(bcmath_number_obj_t), ce);
+
+	zend_object_std_init(&intern->std, ce);
+	object_properties_init(&intern->std, ce);
+
+	intern->num = NULL;
+	intern->value = NULL;
+	intern->scale = 1;
+
+	return &intern->std;
+}
+
+static void bcmath_number_free(zend_object *obj)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	if (intern->num) {
+		bc_free_num(&intern->num);
+		intern->num = NULL;
+	}
+	if (intern->value) {
+		zend_string_release(intern->value);
+		intern->value = NULL;
+	}
+	zend_object_std_dtor(&intern->std);
+}
+
+static zend_object *bcmath_number_clone(zend_object *obj)
+{
+	bcmath_number_obj_t *original = get_bcmath_number_from_obj(obj);
+	bcmath_number_obj_t *clone = get_bcmath_number_from_obj(bcmath_number_create(bcmath_number_ce));
+
+	clone->num = bc_copy_num(original->num);
+	if (original->value) {
+		clone->value = zend_string_copy(original->value);
+	}
+	clone->scale = original->scale;
+
+	return &clone->std;
+}
+
+static HashTable *bcmath_number_get_properties_for(zend_object *obj, zend_prop_purpose purpose)
+{
+	switch (purpose) {
+		case ZEND_PROP_PURPOSE_DEBUG:
+		case ZEND_PROP_PURPOSE_SERIALIZE:
+		case ZEND_PROP_PURPOSE_VAR_EXPORT:
+		case ZEND_PROP_PURPOSE_JSON:
+		case ZEND_PROP_PURPOSE_ARRAY_CAST:
+			break;
+		default:
+			return zend_std_get_properties_for(obj, purpose);
+	}
+
+	zval zv;
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	HashTable *props = zend_array_dup(zend_std_get_properties(obj));
+
+	ZVAL_STR_COPY(&zv, bcmath_number_value_to_str(intern));
+	zend_hash_str_update(props, "value", sizeof("value")-1, &zv);
+	ZVAL_LONG(&zv, intern->scale);
+	zend_hash_str_update(props, "scale", sizeof("scale")-1, &zv);
+
+	return props;
+}
+
+static zval *bcmath_number_write_property(zend_object *obj, zend_string *name, zval *value, void **cache_slot)
+{
+	if (zend_string_equals_literal(name, "value") || zend_string_equals_literal(name, "scale")) {
+		zend_throw_error(NULL, "Cannot modify readonly property %s::$%s", ZSTR_VAL(obj->ce->name), ZSTR_VAL(name));
+		return &EG(error_zval);
+	}
+
+	zend_throw_error(NULL, "Cannot create dynamic property %s::$%s", ZSTR_VAL(obj->ce->name), ZSTR_VAL(name));
+	return &EG(error_zval);
+}
+
+static zval *bcmath_number_read_property(zend_object *obj, zend_string *name, int type, void **cache_slot, zval *rv)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+
+	if (zend_string_equals_literal(name, "value")) {
+		ZVAL_STR_COPY(rv, bcmath_number_value_to_str(intern));
+		return rv;
+	}
+
+	if (zend_string_equals_literal(name, "scale")) {
+		ZVAL_LONG(rv, intern->scale);
+		return rv;
+	}
+
+	return zend_std_read_property(obj, name, type, cache_slot, rv);
+}
+
+static int bcmath_number_isset_property(zend_object *obj, zend_string *name, int check_empty, void **cache_slot)
+{
+	return zend_string_equals_literal(name, "value") || zend_string_equals_literal(name, "scale");
```

Comment:
This implementation won't work properly for when you do `isset($number->value)` where `value` is equal to the string `"0"`.

Reply:
Does that mean it won't have the same behavior as `isset($a)`? ($a = '0')

Reply:
Meant to say "not equal". Note the `check_empty` argument here. `bcmath_number_isset_property` will be called for both isset and empty.
`empty(new Number('0')->value)` should return true (as it does now), but `empty(new Number('1')->value)` should return false, but it doesn't because it's using the same implementation as isset.

Reply:
Thanks, understand.
Fixed in https://github.com/php/php-src/pull/13741/commits/4acfe22e8b201fd660b9510b95e1349c37188632